### ISSUE TITLE
Support floating hint for tiling window managers

### ISSUE THREAD

**Issue by jamadazi** [rust-windowing/winit#862]
On X11 (and I think Wayland too?) there is a standard to provide extra hints to the window manager when creating a window. One of the things this allows is requesting that a tiling/dynamic window manager not tile the window (open it in floating mode instead).

This is an important feature in order to provide a good user experience for people using such window managers, as there are many kinds of application windows which should be opened this way (games should work this way when not fullscreen and normal GUI apps should use this for dialogs, settings windows, floating toolbars, etc.).

AFAIK `winit` does not currently support this. Please consider adding support.

**Comment by SlavojZizekIsASnack** [comment]
i3 determines if a window should be floating by the [_NET_WM_WINDOW_TYPE](https://standards.freedesktop.org/wm-spec/wm-spec-latest.html#idm140200472629520). It is my understanding that the following variants are floats by default:
- dialog, 
- utility, 
- toolbar, 
- splash

**Comment by vberger** [comment]
> (and I think Wayland too?)

I'm not aware of any relatively standard mechanism for doing so, sadly.

**Comment by ElArtista** [comment]
This is how GLFW does it: https://github.com/glfw/glfw/blob/1adfbde4d7fb862bb36d4a20e05d16bf712170f3/src/x11_window.c#L2635

**Comment by ardijanr** [comment]
Setting fixed window size and making it not resizable seems to force floating window on i3 and dwm. Though there should be some standard way to do this.

**Comment by ElArtista** [comment]
Can confirm

**Comment by akvadrako** [comment]
> Setting fixed window size and making it not resizable seems to force floating window on i3 and dwm. Some standard would be nice.

This also works on sway; one can also enable resizing after it's drawn. If someone is interested in getting a standard feature it has been suggested to make a PR in the [wayland-protocols](https://gitlab.freedesktop.org/wayland/wayland-protocols) repo.

**Comment by kchibisov** [comment]
You on wayland control the size yourself, you can resize however you want, it's just entirely up to you to handle all of that.

**Comment by vimpostor** [comment]
> you can resize however you want, it's just entirely up to you to handle all of that.

Right, if you use non-resizable mode, you can effectively simulate support for resizing by adding your own mouse handlers that always update `maximumWidth == minimumWidth` and `maximumHeight == minimumHeight` to some new value.

That is a ridiculous workaround though and would mean reimplementing basic resizing logic just to be able to have a window be floating by default on sway, while also being resizable.

**Comment by akvadrako** [comment]
Another workaround is to initially set the window to fixed size. Sway will make it float. After it displays (under a second later) you can make it resizable. I've done this before and it works, though it's pretty hacky.

**Comment by kchibisov** [comment]
What's the end goal here btw? If to show dialog, there's a new protocol https://wayland.app/protocols/xdg-dialog-v1 for them which we may utilize.

**Comment by vimpostor** [comment]
> What's the end goal here btw?

What the issue description above says, i.e. float a window.

> there's a new protocol https://wayland.app/protocols/xdg-dialog-v1

It's quite limited and doesn't work for top-level windows. Some windows for system utilities might want to be floating by default, but they are not a dialog so your protocol is no use for them.
